queue/driver_mongo.go

```diff
@@ -573,26 +576,102 @@ func (d *mongoDriver) getAtomicQuery(jobName string, modCount int) bson.M {
 }
 
 func isMongoDupKey(err error) bool {
-	// TODO  (EVG-13591): handle duplicate write errors due to scopes separately
-	// from write errors due to job ID clashes.
-	we, ok := errors.Cause(err).(mongo.WriteException)
-	if !ok {
-		return false
+	dupKeyErrs := getMongoDupKeyErrors(err)
+	return dupKeyErrs.writeConcernError != nil || len(dupKeyErrs.writeErrors) != 0 || dupKeyErrs.commandError != nil
+}
+
+func isMongoDupScope(err error) bool {
+	dupKeyErrs := getMongoDupKeyErrors(err)
+	if wce := dupKeyErrs.writeConcernError; wce != nil {
+		if strings.Contains(wce.Message, " scopes_1 ") {
+			return true
+		}
 	}
-	if we.WriteConcernError != nil {
-		wce := we.WriteConcernError
-		return wce.Code == 11000 || wce.Code == 11001 || wce.Code == 12582 || wce.Code == 16460 && strings.Contains(wce.Message, " E11000 ")
+
+	for _, werr := range dupKeyErrs.writeErrors {
+		if strings.Contains(werr.Message, " scopes_1 ") {
+			return true
+		}
 	}
-	if we.WriteErrors != nil && len(we.WriteErrors) > 0 {
-		for _, wErr := range we.WriteErrors {
-			if wErr.Code == 11000 {
-				return true
-			}
+
+	if ce := dupKeyErrs.commandError; ce != nil {
+		if strings.Contains(ce.Message, " scopes_1 ") {
+			return true
 		}
 	}
+
 	return false
 }
 
+type mongoDupKeyErrors struct {
+	writeConcernError *mongo.WriteConcernError
+	writeErrors       []mongo.WriteError
+	commandError      *mongo.CommandError
+}
+
+func getMongoDupKeyErrors(err error) mongoDupKeyErrors {
+	var dupKeyErrs mongoDupKeyErrors
+
+	if we, ok := errors.Cause(err).(mongo.WriteException); ok {
+		dupKeyErrs.writeConcernError = getMongoDupKeyWriteConcernError(we)
+		dupKeyErrs.writeErrors = getMongoDupKeyWriteErrors(we)
+	}
+
+	if ce, ok := errors.Cause(err).(mongo.CommandError); ok {
+		dupKeyErrs.commandError = getMongoDupKeyCommandError(ce)
+	}
+
+	return dupKeyErrs
+}
+
+func getMongoDupKeyWriteConcernError(err mongo.WriteException) *mongo.WriteConcernError {
+	wce := err.WriteConcernError
+	if wce == nil {
+		return nil
+	}
+
+	switch wce.Code {
+	case 11000, 11001, 12582:
```

Comment:
Can we put comments on what these codes mean and why we ignore some?

Reply:
This logic has existed already, I just moved it into a switch. This appears to be [copy-pasted from the go driver](https://github.com/mongodb/mongo-go-driver/blob/213fb80b373f70dba4f9f516dc4c718abe41c76b/mongo/errors.go#L87-L96). When we upgrade the go driver to the latest, we could potentially replace this.

Reply:
Is there a reason we can't update the go driver now?

Reply:
Mostly I don't want to deal with the possibility that it doesn't work for some subtle reason and also that I don't want to revendor the driver in Barque/Cedar/Evergreen just for a tiny piece of code. I also have some concerns that it might not work the same as it did before because the duplicate key error checks differ between server versions. We can deal with upgrading the Go driver to the latest version later.